vm/systemSmartContracts/auction.go

```diff
@@ -0,0 +1,694 @@
+	ownerAddress := s.eei.GetStorage([]byte(ownerKey))
+	if ownerAddress != nil {
+		log.Error("smart contract was already initialized")
+		return vmcommon.UserError
+	}
+
+	s.eei.SetStorage([]byte(ownerKey), args.CallerAddr)
+	return vmcommon.Ok
+}
+
+func (s *stakingAuctionSC) verifyIfKeysExist(registeredKeys [][]byte, arguments [][]byte, maxNumNodes uint64) ([][]byte, error) {
+	registeredKeysMap := make(map[string]struct{})
+
+	for _, blsKey := range registeredKeys {
+		registeredKeysMap[string(blsKey)] = struct{}{}
+	}
+
+	newKeys := make([][]byte, 0)
+	for i := uint64(1); i < maxNumNodes+1; i++ {
+		if _, ok := registeredKeysMap[string(arguments[i])]; ok {
+			continue
+		}
+
+		newKeys = append(newKeys, arguments[i])
+	}
+
+	for _, newKey := range newKeys {
+		data := s.eei.GetStorage(newKey)
+		if len(data) > 0 {
+			return nil, vm.ErrKeyAlreadyRegistered
+		}
+	}
+
+	return newKeys, nil
+}
+
+func (s *stakingAuctionSC) stake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	config := s.getConfig(s.eei.BlockChainHook().CurrentEpoch())
+
+	registrationData, err := s.getRegistrationData(args.CallerAddr)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	registrationData.TotalStakeValue.Add(registrationData.TotalStakeValue, args.CallValue)
+	if registrationData.TotalStakeValue.Cmp(config.MinStakeValue) < 0 || args.CallValue.Sign() <= 0 {
+		return vmcommon.UserError
+	}
+
+	lenArgs := len(args.Arguments)
+	if lenArgs == 0 {
+		if len(registrationData.BlsPubKeys) == 0 {
+			log.Debug("not enough arguments to process stake function")
+			return vmcommon.UserError
+		}
+
+		err := s.saveRegistrationData(args.CallerAddr, registrationData)
+		if err != nil {
+			return vmcommon.UserError
+		}
+
+		return vmcommon.Ok
+	}
+
+	maxNodesToRun := big.NewInt(0).SetBytes(args.Arguments[0]).Uint64()
+	if uint64(lenArgs) < maxNodesToRun+1 {
+		log.Debug("not enough arguments to process stake function")
+		return vmcommon.UserError
+	}
+
+	newKeys, err := s.verifyIfKeysExist(registrationData.BlsPubKeys, args.Arguments, maxNodesToRun)
+	if err != nil {
+		log.Debug("Staking with already existing key is not allowed")
+		return vmcommon.UserError
+	}
+
+	for _, blsKey := range newKeys {
+		_, err := s.kg.PublicKeyFromByteArray(blsKey)
+		if err != nil {
+			log.Debug("bls key is not valid")
+			return vmcommon.UserError
+		}
+
+		s.eei.SetStorage(blsKey, []byte("registered"))
+		registrationData.BlsPubKeys = append(registrationData.BlsPubKeys, blsKey)
+	}
+
+	registrationData.RewardAddress = args.CallerAddr
+	registrationData.MaxStakePerNode = big.NewInt(0).Set(registrationData.TotalStakeValue)
+	registrationData.Epoch = s.eei.BlockChainHook().CurrentEpoch()
+
+	// do the optionals - rewardAddress and maxStakePerNode
+	if uint64(lenArgs) > maxNodesToRun+1 {
+		for i := maxNodesToRun + 1; i < uint64(lenArgs); i++ {
+			if len(args.Arguments[i]) == len(args.CallerAddr) {
+				registrationData.RewardAddress = args.Arguments[i]
```

Comment:
it would be hard for delegation smart contracts to send 2 transactions and watch for the output of those 2. Reward address is only for expert use - delegators in general, if they want in a new address the rewards, for the normal user it shouldn't be used.

Reply:
I think we can implement this in a future PR. Everyone will monitor if their staking transaction got through.

Reply:
ok. will add verification that MaxStakePerNode should not be higher than totalStake.